common/lib/dependabot/pull_request_updater/azure.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+require "dependabot/clients/azure"
+
+module Dependabot
+  class PullRequestUpdater
+    class Azure
+      OBJECT_ID_FOR_BRANCH_DELETE = "0000000000000000000000000000000000000000"
+
+      attr_reader :source, :files, :base_commit, :old_commit, :credentials,
+                  :pull_request_number
+
+      def initialize(source:, files:, base_commit:, old_commit:,
+                     credentials:, pull_request_number:)
+        @source = source
+        @files = files
+        @base_commit = base_commit
+        @old_commit = old_commit
+        @credentials = credentials
+        @pull_request_number = pull_request_number
+      end
+
+      def update
+        return unless pull_request_exists? && source_branch_exists?
+
+        update_source_branch
+      end
+
+      private
+
+      def azure_client_for_source
+        @azure_client_for_source ||=
+          Dependabot::Clients::Azure.for_source(
+            source: source,
+            credentials: credentials
+          )
+      end
+
+      def pull_request_exists?
+        pull_request
+      rescue Dependabot::Clients::Azure::NotFound
+        false
+      end
+
+      def source_branch_exists?
+        azure_client_for_source.branch(source_branch_name)
+      rescue Dependabot::Clients::Azure::NotFound
+        false
+      end
+
+      # Currently the PR diff in ADO shows difference in commits instead of actual diff in files.
+      # This workaround is done to get the target branch commit history on the source branch alongwith file changes
+      def update_source_branch
+        # 1) Push the file changes to a newly created temporary branch (from base commit)
+        new_commit = create_temp_branch
+        # 2) Update PR source branch to point to the temp branch head commit.
+        update_branch(source_branch_name, old_source_branch_commit, new_commit)
+        # 3) Delete temp branch
+        update_branch(temp_branch_name, new_commit, OBJECT_ID_FOR_BRANCH_DELETE)
+      end
+
+      def pull_request
+        @pull_request ||=
+          azure_client_for_source.fetch_pull_request(pull_request_number.to_s)
+      end
+
+      def source_branch_name
+        @source_branch_name ||= pull_request&.fetch("sourceRefName")&.gsub("refs/heads/", "")
+      end
+
+      def create_temp_branch
+        response = azure_client_for_source.create_commit(
+          temp_branch_name,
+          base_commit,
+          commit_message,
+          files,
+          nil
+        )
+
+        JSON.parse(response.body).fetch("refUpdates").first.fetch("newObjectId")
+      end
+
+      def temp_branch_name
+        "#{source_branch_name}-temp"
```

Comment:
Would it be worth generating a short id, e.g. `#{source_branch_name}-temp-Digest::SHA1.hexdigest(source_branch_name)[0..6]`?

Thinking of the case where two jobs try to update the same PR or someone has stashed some changes in a `-temp` version of the current branch, in which case it would be deleted. Seems unlikely but have a feeling I've created branches with temp suffixes in the past when wrangling a rebase for example.

Reply:
Feeling your thoughts here as potential across the spectrum in various ways. You probably have a magic hat full of solutions... let me know if I can help with checklists/more specific contribution.... It's a lot for sure. Extra atd (attn to detail) is always a wise investment.

Reply:
@feelepxyz Digest::SHA1.hexdigest(source_branch_name) produces consistent hashing. So for the case where we have two jobs trying to update the same PR, this might still result in issues. I was thinking of using uuid ([documentation](https://www.rubydoc.info/gems/uuid/2.3.1/UUID)) for uniquely identifying the temp branch. Thoughts?

Reply:
Ah yes you're right, SHA1 is stable, uuid seems like a good candidate 👍

Reply:
Cool made an update accordingly